common-http/src/main/java/io/cdap/common/http/HttpRequests.java

```diff
@@ -103,31 +102,45 @@ public static HttpResponse execute(HttpRequest request, HttpRequestConfig reques
 
     conn.connect();
 
-    try {
-      if (bodySrc != null) {
-        try (InputStream input = bodySrc.getInput(); OutputStream os = conn.getOutputStream()) {
-          ByteStreams.copy(input, os);
-        }
-      }
 
-      try {
-        if (isSuccessful(conn.getResponseCode())) {
-          try (InputStream inputStream = conn.getInputStream()) {
-            return new HttpResponse(conn.getResponseCode(), conn.getResponseMessage(),
-                                    ByteStreams.toByteArray(inputStream), conn.getHeaderFields());
-          }
-        }
-      } catch (FileNotFoundException e) {
-        // Server returns 404. Hence handle as error flow below. Intentional having empty catch block.
+    if (bodySrc != null) {
+      try (InputStream input = bodySrc.getInput(); OutputStream os = conn.getOutputStream()) {
+        ByteStreams.copy(input, os);
       }
+    }
+
+    if (isSuccessful(conn.getResponseCode())) {
+      return new HttpResponse(conn.getResponseCode(), conn.getResponseMessage(),
+                              conn.getInputStream(),
+                              () -> {
```

Comment:
Shouldn't pass a supplier like this, especially the conn.getInputStream() is also passed to the HttpResponse. You should have the reading whole content logic resides in the HttpResponse class instead.

Reply:
I'm now passing HttpURLConnection as discussed offline